infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java

```diff
@@ -0,0 +1,139 @@
+        this.stateTransferBatchProcessor = stateTransferBatchProcessor;
+    }
+
+    private CompletableFuture<Void> handleNonTransferException(CompletableFuture<Void> futures,
+                                                               Throwable ex) {
+        CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+        failedFuture.completeExceptionally(ex);
+        return CFUtils.allOfOrTerminateExceptionally(futures, failedFuture);
+    }
+
+    private CompletableFuture<Void> handleBatchRequest(TransferBatchRequest request,
+                                                       CompletableFuture<Void> allFutures,
+                                                       Semaphore semaphore) throws InterruptedException {
+        // If the request is of type SEGMENT_INIT - it signals to the batch processor that the
+        // consecutive DATA requests all belong to the same segment. Since all the DATA requests
+        // that belong to the same segment have the same number of source nodes, we can dynamically
+        // configure the semaphore to support the average parallelism level for the entire segment
+        // which is equal to the NUM_REQUEST_PER_NODE * number of source nodes in the segment.
+        if (request.getBatchType() == SEGMENT_INIT) {
+            int numPermits = request.getDestinationNodes()
+                    .map(AbstractCollection::size).orElse(1) *
+                    NUM_REQUESTS_PER_NODE;
+            if (numPermits == 0) {
+                throw new IllegalStateException("Number of permits should be equal to" +
+                        " at least one.");
+            }
+            log.trace("Starting a new segment transfer. Parallelism level: {}", numPermits);
+            // Acquire all the permits if any.
+            semaphore.drainPermits();
+            // Release numPermits for the batch transfer futures to acquire.
+            semaphore.release(numPermits);
+        }
+        else if (request.getBatchType() == DATA) {
+            semaphore.acquire();
+            CompletableFuture<Void> batchTransferResult =
+                    stateTransferBatchProcessor
+                            .transfer(request)
+                            .thenApply(response -> {
+                                semaphore.release();
+                                if (response.getStatus() == SUCCEEDED) {
+                                    return null;
+                                } else if (response.getStatus() == FAILED &&
+                                        response.getCauseOfFailure().isPresent()) {
+                                    throw response.getCauseOfFailure().get();
+                                } else {
+                                    throw new StateTransferBatchProcessorException();
+                                }
+                            });
+
+            allFutures = CFUtils.allOfOrTerminateExceptionally(allFutures, batchTransferResult);
```

Comment:
I think this still does not solve the problem we have.
Here is your impl of CFUtils.allOfOrTerminateExceptionally:
```java
CompletableFuture<Void> result = CompletableFuture.allOf(futures);
for (CompletableFuture<?> future : futures) {
    future.handle((res, ex) -> ex == null || result.completeExceptionally(ex));
}
```

Let's suppose this is the current allFutures structure:
`a3<a2<a1<f0, f1>, f2>, f3>`
where a_i means allFuture, f_i means the individual transfer future.
If f1 fails at this moment, it will cancel a1, however f2 and f3 are not finished yet, so even though a1 is exceptionally completedly, the curerrnt allFuture, i.e. a3 still waits until f2 and f3 finishes. So if there are more futures, we won't be able to jump out of the loop, right?

Reply:
@WenbinZhu I am pretty sure it works fine and I also wrote a test for it. You can check the test testRunParallelTransferShortCircuitsOnFailure. You can replace the CFUtils.allOfOrTerminateExceptionally with CompletableFuture.allOf and see that the test does not go through.

Reply:
@WenbinZhu Let me know if you think the test does not capture it.

Reply:
Ok I see, I didn't notice both futures in allOf are attached a handler, so it can exit. But the problem is when exception happens in the lowest level future, the exception needs to propagate multiple levels all the way to the outermost future, while we could have a variable to jump out and only fail the outmost future once. But I think it's ok for now.